src/stream.rs

```diff
@@ -49,6 +52,12 @@ const SEND_BUFFER_SIZE: usize = 5;
 #[cfg(not(test))]
 const SEND_BUFFER_SIZE: usize = 4096;
 
+// The default size of the receiver stream flow control window.
+const DEFAULT_STREAM_WINDOW: u64 = 32 * 1024;
```

Comment:
Does this default override the configuration value set by the application in e.g. `set_initial_max_stream_data_bidi_local()` and `set_initial_max_stream_data_bidi_remote()`?

Maybe instead we need to take into account those values. And that would mean we need separate flow control behaviour for bidi streams initiated locally, bidi streams initiated remotely, and uni stream initiated remotely.

Reply:
DEFAULT_{CONNECTION,STREAM}_WINDOW is a different value from initial stream/connection max_data. in the flowcontrol.rs there is `window` variable is limited by those default values and (so, max_data will be consumed + window). Since those are hardcoded, it's a good idea to make it configurable, but don't be confused with max_data transport parameter which defines initial limit only and nothing to do with window size.

Reply:
I added `config.set_max_connection_window()` and `config.set_max_stream_window()` API to set the max window size of the connection and the stream.

Reply:
I'm still confused. In the code we have `cmp::min(max_data, DEFAULT_STREAM_WINDOW)`, which means tha if the application sets a lower initial stream data value (i.e. `max_data`) than the "default" value and a lower max window value, we will just ignore the application-set values and simply use `DEFAULT_STREAM_WINDOW` as the initial window size. And later, when autotune is done, the new window will just be `max_window`.

So it's still not very clear to me what the point of this constant is. Why not just use `max_data` without doing the `cmp::min()`?

Reply:
And while we are at it, why even do autotuning at all, and not just use `max_window` from the start?

Reply:
Actually, I just realized I was probably confusing min with max :facepalm: ...

Reply:
`cmp::min()` is used for flow control window initialization when the max_data (whether it's connection or stream) is too small (e.g. tests) in such case, having a small window will make sense and it's helpful for passing most of existing tests without a large modification (I only changed 2 tests). In practive it doesn't have much meaning

Reply:
> And while we are at it, why even do autotuning at all, and not just use `max_window` from the start?

Interesting. I think the reason to have a flow control window is to properly manage receiver buffer size. If it's too high each connection/stream might need more memory even for a small BDP connection and MAX_DATA update might not be delivered in time.

I see some of implementation use a very high max_data in the beginning (e.g. h2load + quic's initial max_data is  2^30-1 and window size is 2^26-1, but it's a test tool). In practical use I believe it's better to tune the window size dynamically.